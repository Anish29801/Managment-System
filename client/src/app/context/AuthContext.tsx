"use client";

import { createContext, useContext, useState, useEffect, ReactNode } from "react";

interface User {
  name: string;
  email: string;
}

interface AuthContextType {
  user: User | null;
  setUser: (user: User | null) => void;
  logout: () => void;
}

const AuthContext = createContext<AuthContextType | undefined>(undefined);

export const AuthProvider = ({ children }: { children: ReactNode }) => {
  const [user, setUser] = useState<User | null>(null);

  // Fetch user on mount
  useEffect(() => {
    const fetchUser = async () => {
      const token = localStorage.getItem("token");
      if (!token) return;

      try {
        const res = await fetch("http://localhost:8000/users/me", {
          headers: { Authorization: `Bearer ${token}` },
        });
        if (res.ok) {
          const data = await res.json();
          setUser(data.user);
        } else {
          setUser(null);
        }
      } catch {
        setUser(null);
      }
    };

    fetchUser();
  }, []);

  const logout = () => {
    localStorage.removeItem("token");
    setUser(null);
  };

  return (
    <AuthContext.Provider value={{ user, setUser, logout }}>
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) throw new Error("useAuth must be used within AuthProvider");
  return context;
};
