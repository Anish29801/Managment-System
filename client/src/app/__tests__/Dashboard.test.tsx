import React from "react";
import { render, screen, waitFor, act } from "@testing-library/react";
import Dashboard from "@/app/page";
import axios from "axios";
import { useAuth } from "@/app/context/AuthContext";


// 🧩 Mock Clock component for stable snapshot and time-independent tests
jest.mock("@/app/components/Clock", () => ({
  __esModule: true,
  default: () => <div data-testid="clock">Mock Clock</div>,
}));

// 🧩 Mock useAuth hook to simulate logged in/out user
jest.mock("@/app/context/AuthContext", () => ({
  useAuth: jest.fn(),
}));

// 🧩 Mock Axios to prevent real HTTP requests
jest.mock("axios");
const mockAxios = axios as jest.Mocked<typeof axios>;

describe("Dashboard component", () => {
  const mockUser = { name: "Anish" };

  beforeEach(() => {
    jest.clearAllMocks(); // Clear previous mocks
  });

  // ----------------------
  // ✅ 1. Shows HeroSection when user is not logged in
  it("shows HeroSection when user is not logged in", async () => {
    (useAuth as jest.Mock).mockReturnValue({ user: null });

    await act(async () => {
      render(<Dashboard />);
    });

    expect(screen.getByText(/Sign Up/i)).toBeInTheDocument();
    expect(screen.getByText(/Login/i)).toBeInTheDocument();
  });

  // ----------------------
  // ✅ 2. Fetches and displays tasks
  it("fetches and displays tasks", async () => {
    (useAuth as jest.Mock).mockReturnValue({ user: mockUser });

    // 🧩 Mock Axios response for tasks
    mockAxios.get.mockResolvedValueOnce({
      data: [
        { id: 1, title: "Mock Task 1", completed: false },
        { id: 2, title: "Mock Task 2", completed: true },
      ],
    });

    await act(async () => {
      render(<Dashboard />);
    });

    // 🧩 Wait for tasks to appear
    expect(await screen.findByText(/Mock Task 1/i)).toBeInTheDocument();
    expect(await screen.findByText(/Mock Task 2/i)).toBeInTheDocument();
  });

  // ----------------------
  // ✅ 3. Shows Clock component
  it("renders Clock component", async () => {
    (useAuth as jest.Mock).mockReturnValue({ user: mockUser });

    await act(async () => {
      render(<Dashboard />);
    });

    expect(screen.getByTestId("clock")).toBeInTheDocument();
    expect(screen.getByText("Mock Clock")).toBeInTheDocument();
  });

  // ----------------------
  // ✅ 4. Matches snapshot when logged in
  it("matches snapshot when logged in", async () => {
    (useAuth as jest.Mock).mockReturnValue({ user: mockUser });

    // 🧩 Mock Axios for consistent task list
    mockAxios.get.mockResolvedValueOnce({
      data: [
        { id: 1, title: "Mock Task 1", completed: false },
        { id: 2, title: "Mock Task 2", completed: true },
      ],
    });

    let asFragment: () => DocumentFragment;
    await act(async () => {
      const renderResult = render(<Dashboard />);
      asFragment = renderResult.asFragment;
    });

    await waitFor(() => screen.getByText(/Hi, Anish Good/i));
    expect(asFragment!()).toMatchSnapshot(); // stable snapshot with mocked Clock and tasks
  });

  // ----------------------
  // ✅ 5. Handles empty tasks gracefully
  it("shows 'No tasks here.' when task list is empty", async () => {
    (useAuth as jest.Mock).mockReturnValue({ user: mockUser });

    mockAxios.get.mockResolvedValueOnce({ data: [] });

    await act(async () => {
      render(<Dashboard />);
    });

    expect(await screen.findAllByText(/No tasks here/i)).toHaveLength(3); // pending, in-progress, completed
  });

  // ----------------------
  // ✅ 6. Handles Axios error gracefully
  it("shows toast message on fetch error", async () => {
    (useAuth as jest.Mock).mockReturnValue({ user: mockUser });

    mockAxios.get.mockRejectedValueOnce(new Error("Network Error"));

    await act(async () => {
      render(<Dashboard />);
    });

    expect(await screen.findByText(/❌ Failed to fetch tasks/i)).toBeInTheDocument();
  });

  // Additional tests like Add Task button click, search, filter can go here
});
