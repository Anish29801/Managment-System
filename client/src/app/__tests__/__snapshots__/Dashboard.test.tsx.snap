// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Dashboard component matches snapshot when logged in 1`] = `
<DocumentFragment>
  <div
    class="min-h-screen p-6 flex flex-col relative text-white bg-gradient-to-br from-gray-900 via-gray-800 to-black overflow-hidden"
  >
    <div
      class="absolute inset-0 bg-white/5 backdrop-blur-sm pointer-events-none z-0"
    />
    <div
      class="relative z-10 w-full"
    >
      <div
        class="flex flex-col md:flex-row justify-between items-start md:items-center mb-6 gap-6"
      >
        <div
          class="flex flex-col md:flex-row items-start md:items-center gap-4 w-full md:w-auto"
        >
          <h1
            class="text-3xl font-bold text-white tracking-wide"
          >
            Hi, Anish Good Afternoon
          </h1>
          <div
            class="text-white font-semibold text-base sm:text-lg md:text-xl lg:text-2xl"
          >
            5:27 PM
          </div>
        </div>
        <div
          class="grid grid-cols-1 sm:grid-cols-2 lg:flex gap-4 w-full bg-gray-800/70 p-4 rounded-xl shadow-lg"
        >
          <div
            class="relative w-full"
          >
            <input
              class="w-full pl-10 pr-4 py-2 rounded-lg bg-gray-700 text-white placeholder-gray-400 focus:outline-none focus:ring-2 focus:ring-blue-500"
              placeholder="Search tasks..."
              type="text"
              value=""
            />
            <svg
              class="absolute left-3 top-2.5 text-gray-400"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
              />
            </svg>
          </div>
          <div
            class="flex flex-col sm:flex-row gap-2 w-full items-center"
          >
            <input
              class="w-full px-3 py-2 pr-8 rounded-lg bg-gray-700 text-white placeholder-gray-400 focus:outline-none focus:ring-2 focus:ring-blue-500 appearance-none"
              type="date"
              value=""
            />
            <div
              class="flex justify-center items-center w-full sm:w-auto"
            >
              <svg
                class="text-gray-300 text-lg block sm:hidden"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                />
              </svg>
              <svg
                class="text-gray-300 text-lg hidden sm:block"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 320 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
                />
              </svg>
            </div>
            <input
              class="w-full px-3 py-2 pr-8 rounded-lg bg-gray-700 text-white placeholder-gray-400 focus:outline-none focus:ring-2 focus:ring-blue-500 appearance-none"
              type="date"
              value=""
            />
          </div>
          <div
            class="w-full sm:w-auto"
          >
            <button
              class="w-full h-10 px-4 rounded-md bg-blue-600 hover:bg-blue-700 shadow-md transition-colors text-sm font-medium"
            >
              Add Task
            </button>
          </div>
        </div>
      </div>
      <div
        class="flex flex-col md:flex-row gap-6 mt-6"
      >
        <div
          class="w-full sm:w-full md:w-1/2 lg:w-1/3 p-2"
        >
          <div
            class="bg-gray-800 rounded-2xl shadow-xl p-4 pt-8 relative flex flex-col"
          >
            <h2
              class="absolute -top-3 left-4 bg-gray-900 px-3 py-1 text-white font-bold rounded-full text-sm shadow"
            >
              Pending
            </h2>
            <ul
              class="space-y-4 min-h-[200px] mt-2 overflow-y-auto max-h-[70vh] sm:max-h-[60vh] md:max-h-[65vh]"
              data-rfd-droppable-context-id="«rf»"
              data-rfd-droppable-id="pending"
            >
              <p
                class="text-gray-400 text-sm"
              >
                No tasks here.
              </p>
            </ul>
          </div>
        </div>
        <div
          class="w-full sm:w-full md:w-1/2 lg:w-1/3 p-2"
        >
          <div
            class="bg-gray-800 rounded-2xl shadow-xl p-4 pt-8 relative flex flex-col"
          >
            <h2
              class="absolute -top-3 left-4 bg-gray-900 px-3 py-1 text-white font-bold rounded-full text-sm shadow"
            >
              In Progress
            </h2>
            <ul
              class="space-y-4 min-h-[200px] mt-2 overflow-y-auto max-h-[70vh] sm:max-h-[60vh] md:max-h-[65vh]"
              data-rfd-droppable-context-id="«rf»"
              data-rfd-droppable-id="inprogress"
            >
              <p
                class="text-gray-400 text-sm"
              >
                No tasks here.
              </p>
            </ul>
          </div>
        </div>
        <div
          class="w-full sm:w-full md:w-1/2 lg:w-1/3 p-2"
        >
          <div
            class="bg-gray-800 rounded-2xl shadow-xl p-4 pt-8 relative flex flex-col"
          >
            <h2
              class="absolute -top-3 left-4 bg-gray-900 px-3 py-1 text-white font-bold rounded-full text-sm shadow"
            >
              Completed
            </h2>
            <ul
              class="space-y-4 min-h-[200px] mt-2 overflow-y-auto max-h-[70vh] sm:max-h-[60vh] md:max-h-[65vh]"
              data-rfd-droppable-context-id="«rf»"
              data-rfd-droppable-id="completed"
            >
              <p
                class="text-gray-400 text-sm"
              >
                No tasks here.
              </p>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
